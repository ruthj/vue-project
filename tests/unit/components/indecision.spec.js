import {shallowMount} from '@vue/test-utils'
import Indecision from '@/components/Indecision'


describe('Indecision component', () => {
    let wrapper 
    let clgSpy
    let getAnwerSpy

    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision); //mounts the component no the app.js
        clgSpy = jest.spyOn(console, 'log')
        getAnwerSpy = jest.spyOn(wrapper.vm, 'getAnswer')//vm VUE instance
        jest.clearAllMocks()
    })

    it('Should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Write in the input should not trigger anything (console.log) ', async() => {
        
        //const getAnwerSpy = jest.spyOn(wrapper.vm, 'getAnswer')//vm VUE instance
        const input = wrapper.find('input')
        await input.setValue('Hi world')

        expect(clgSpy).toBeCalledTimes(1)
        expect(getAnwerSpy).toBeCalledTimes(0)
        //OR
        //expect(getAnwerSpy).not.toBeCalledTimes()
    })

    test('Write question mark should trigger fetch ', async() => {
       // const getAnwerSpy = jest.spyOn(wrapper.vm, 'getAnswer')//vm VUE instance
        const input = wrapper.find('input')
        await input.setValue('Hi world?')

       // expect(clgSpy).toBeCalledTimes(1)
        expect(getAnwerSpy).toBeCalledTimes(1)
    })

    test('test getAnswer() ', async() => {
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Si!')
    })

    test('test getAnswer() wrong way', async() => {
        
        fetch.mockImplementationOnce( () => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('Couldnt load from API!')
    })

})