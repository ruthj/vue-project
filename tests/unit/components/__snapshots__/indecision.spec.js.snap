// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component Should match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input ype="text" placeholder="Hazme una pregunta">
  <p>Recueda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
